File: packages/core/src/engine/accrual.ts
import { assertU128, assertU64, clampBig, maxBig } from '../math/fixed';
import { ErrorCode, fail } from './errors';
import type { PositionState, SideState } from './types';

/**
 * Time-weighted conviction accumulators (ECONOMICS §5).
 *
 * Every function is pure, bigint-only, and written so it can be transcribed
 * line-by-line into the Anchor program. Timestamps are clamped into
 * `[startTs, endTs]` before differencing, so no time outside the Arena window
 * is ever counted and repeated accrual at the same timestamp is a no-op.
 */

export interface Window {
  startTs: bigint;
  endTs: bigint;
}

/** Clamp a timestamp into the accrual window. */
export function clampTs(ts: bigint, w: Window): bigint {
  return clampBig(ts, w.startTs, w.endTs);
}

/** Seconds of accrual between two raw timestamps, inside the window only. */
export function accrualDelta(lastTs: bigint, now: bigint, w: Window): bigint {
  if (now < lastTs) fail(ErrorCode.NonMonotonicClock, `now ${now} < last ${lastTs}`);
  const dt = clampTs(now, w) - clampTs(lastTs, w);
  return dt < 0n ? 0n : dt; // cannot be negative after clamping, defensive
}

/** Accrue a side aggregate by `dt` seconds. */
export function accrueSide(side: SideState, dt: bigint): SideState {
  if (dt === 0n) return side;
  return {
    ...side,
    unitSeconds: assertU128(side.unitSeconds + side.units * dt, 'side.unit_seconds'),
    effUnitSeconds: assertU128(side.effUnitSeconds + side.effUnits * dt, 'side.eff_unit_seconds'),
  };
}

/** Accrue a position up to `now`, updating its touch timestamp. */
export function accruePosition(p: PositionState, now: bigint, w: Window): PositionState {
  const dt = accrualDelta(p.lastTouchTs, now, w);
  return {
    ...p,
    unitSeconds: assertU128(p.unitSeconds + p.units * dt, 'position.unit_seconds'),
    effUnitSeconds: assertU128(p.effUnitSeconds + p.effUnits * dt, 'position.eff_unit_seconds'),
    lastTouchTs: maxBig(p.lastTouchTs, now),
  };
}

export function newPosition(
  owner: string,
  side: PositionState['side'],
  now: bigint,
): PositionState {
  return {
    owner,
    side,
    units: 0n,
    unitSeconds: 0n,
    effUnits: 0n,
    effUnitSeconds: 0n,
    entryTs: now,
    lastTouchTs: now,
    claimed: false,
    deposits: 0n,
    feePaid: 0n,
    forfeited: false,
  };
}

/**
 * Apply a deposit tranche of `units` at multiplier `mQ4` to an already-accrued
 * position and side (ECONOMICS §5.4).
 */
export function applyDeposit(
  p: PositionState,
  side: SideState,
  units: bigint,
  mQ4: bigint,
  feePaid: bigint,
): { position: PositionState; side: SideState } {
  if (units <= 0n) fail(ErrorCode.ZeroAmount);
  assertU64(units, 'units');
  const eff = assertU128(units * mQ4, 'eff_units');
  const isNew = p.units === 0n && p.deposits === 0n;
  return {
    position: {
      ...p,
      units: assertU64(p.units + units, 'position.units'),
      effUnits: assertU128(p.effUnits + eff, 'position.eff_units'),
      deposits: p.deposits + 1n,
      feePaid: p.feePaid + feePaid,
    },
    side: {
      ...side,
      units: assertU64(side.units + units, 'side.units'),
      effUnits: assertU128(side.effUnits + eff, 'side.eff_units'),
      participants: isNew ? side.participants + 1n : side.participants,
    },
  };
}

/**
 * Exit `x` units from an already-accrued position (ECONOMICS §5.5).
 *
 * When `forfeit` is true (now < endTs and Arena live) both live and accrued
 * weight are reduced proportionally; when false (after end_ts, or Arena
 * cancelled/settled) only units move and weight is untouched.
 */
export function applyExit(
  p: PositionState,
  side: SideState,
  x: bigint,
  forfeit: boolean,
): { position: PositionState; side: SideState; removedEff: bigint; removedEffSeconds: bigint } {
  if (x <= 0n) fail(ErrorCode.ZeroAmount);
  if (x > p.units) fail(ErrorCode.InsufficientUnits, `exit ${x} > held ${p.units}`);
  const u = p.units;
  const keepNum = u - x;
  let removedEff = 0n;
  let removedEffSeconds = 0n;
  let removedUnitSeconds = 0n;
  if (forfeit) {
    // removed = total − total × keep / u   (floor keeps the *kept* part rounded down,
    // so the forfeited part absorbs rounding — never favours the exiting user)
    removedEff = p.effUnits - (p.effUnits * keepNum) / u;
    removedEffSeconds = p.effUnitSeconds - (p.effUnitSeconds * keepNum) / u;
    removedUnitSeconds = p.unitSeconds - (p.unitSeconds * keepNum) / u;
  }
  const units = keepNum;
  return {
    position: {
      ...p,
      units,
      unitSeconds: p.unitSeconds - removedUnitSeconds,
      effUnits: p.effUnits - removedEff,
      effUnitSeconds: p.effUnitSeconds - removedEffSeconds,
      forfeited: p.forfeited || (forfeit && units === 0n),
    },
    side: {
      ...side,
      units: side.units - x,
      effUnits: side.effUnits - removedEff,
      effUnitSeconds: side.effUnitSeconds - removedEffSeconds,
    },
    removedEff,
    removedEffSeconds,
  };
}

File: programs/tribe_arena/src/engine/accrual.rs
//! Time-weighted conviction accumulators — transcription of
//! `packages/core/src/engine/accrual.ts` (ECONOMICS §5).

use crate::engine::math::{add_u128, mul_u128, sub_u128};
use crate::errors::TribeError;
use crate::state::{Position, SideState};
use anchor_lang::prelude::*;

#[derive(Clone, Copy, Debug)]
pub struct Window {
    pub start_ts: i64,
    pub end_ts: i64,
}

pub fn clamp_ts(ts: i64, w: &Window) -> i64 {
    ts.clamp(w.start_ts, w.end_ts)
}

/// Seconds of accrual between two raw timestamps, inside the window only.
pub fn accrual_delta(last_ts: i64, now: i64, w: &Window) -> Result<u128> {
    require!(now >= last_ts, TribeError::NonMonotonicClock);
    let dt = clamp_ts(now, w) - clamp_ts(last_ts, w);
    Ok(if dt < 0 { 0 } else { dt as u128 })
}

pub fn accrue_side(side: &mut SideState, dt: u128) -> Result<()> {
    if dt == 0 {
        return Ok(());
    }
    side.unit_seconds = add_u128(side.unit_seconds, mul_u128(side.units as u128, dt)?)?;
    side.reward_unit_seconds =
        add_u128(side.reward_unit_seconds, mul_u128(side.units as u128, dt)?)?;
    side.eff_unit_seconds = add_u128(side.eff_unit_seconds, mul_u128(side.eff_units, dt)?)?;
    Ok(())
}

/// Accrue a position up to `now`, updating its touch timestamp.
pub fn accrue_position(p: &mut Position, now: i64, w: &Window) -> Result<()> {
    let dt = accrual_delta(p.last_touch_ts, now, w)?;
    p.unit_seconds = add_u128(p.unit_seconds, mul_u128(p.units as u128, dt)?)?;
    p.eff_unit_seconds = add_u128(p.eff_unit_seconds, mul_u128(p.eff_units, dt)?)?;
    p.last_touch_ts = p.last_touch_ts.max(now);
    Ok(())
}

/// Apply a deposit tranche of `units` at multiplier `m_q4` (ECONOMICS §5.4).
pub fn apply_deposit(
    p: &mut Position,
    side: &mut SideState,
    units: u64,
    m_q4: u32,
    fee_paid: u64,
) -> Result<()> {
    require!(units > 0, TribeError::ZeroAmount);
    let eff = mul_u128(units as u128, m_q4 as u128)?;
    let is_new = p.units == 0 && p.deposits == 0;
    p.units = p
        .units
        .checked_add(units)
        .ok_or_else(|| error!(TribeError::MathOverflow))?;
    p.eff_units = add_u128(p.eff_units, eff)?;
    p.deposits = p
        .deposits
        .checked_add(1)
        .ok_or_else(|| error!(TribeError::MathOverflow))?;
    p.fee_paid = p
        .fee_paid
        .checked_add(fee_paid)
        .ok_or_else(|| error!(TribeError::MathOverflow))?;
    side.units = side
        .units
        .checked_add(units)
        .ok_or_else(|| error!(TribeError::MathOverflow))?;
    side.eff_units = add_u128(side.eff_units, eff)?;
    if is_new {
        side.participants = side
            .participants
            .checked_add(1)
            .ok_or_else(|| error!(TribeError::MathOverflow))?;
    }
    Ok(())
}

pub struct ExitResult {
    pub removed_eff: u128,
    pub removed_eff_seconds: u128,
    pub removed_unit_seconds: u128,
}

/// Exit `x` units (ECONOMICS §5.5). With `forfeit` both live and accrued
/// weight are reduced proportionally (floor on the kept part); without it
/// only units move.
pub fn apply_exit(
    p: &mut Position,
    side: &mut SideState,
    x: u64,
    forfeit: bool,
) -> Result<ExitResult> {
    require!(x > 0, TribeError::ZeroAmount);
    require!(x <= p.units, TribeError::InsufficientUnits);
    let u = p.units as u128;
    let keep = u - x as u128;
    let mut r = ExitResult {
        removed_eff: 0,
        removed_eff_seconds: 0,
        removed_unit_seconds: 0,
    };
    if forfeit {
        r.removed_eff = sub_u128(p.eff_units, mul_u128(p.eff_units, keep)? / u)?;
        r.removed_eff_seconds =
            sub_u128(p.eff_unit_seconds, mul_u128(p.eff_unit_seconds, keep)? / u)?;
        r.removed_unit_seconds = sub_u128(p.unit_seconds, mul_u128(p.unit_seconds, keep)? / u)?;
    }
    p.units -= x;
    p.unit_seconds = sub_u128(p.unit_seconds, r.removed_unit_seconds)?;
    p.eff_units = sub_u128(p.eff_units, r.removed_eff)?;
    p.eff_unit_seconds = sub_u128(p.eff_unit_seconds, r.removed_eff_seconds)?;
    p.forfeited = p.forfeited || (forfeit && p.units == 0);
    side.units = side
        .units
        .checked_sub(x)
        .ok_or_else(|| error!(TribeError::MathOverflow))?;
    side.reward_unit_seconds = sub_u128(side.reward_unit_seconds, r.removed_unit_seconds)?;
    side.eff_units = sub_u128(side.eff_units, r.removed_eff)?;
    side.eff_unit_seconds = sub_u128(side.eff_unit_seconds, r.removed_eff_seconds)?;
    Ok(r)
}
